File: year2/semester2/mpp/c#/TripClientServerGrpc/Common/Net/ProtoUtils.cs
using System;
using System.Collections.Generic;
using System.Text;
using tripsGrpc;
using Account = TripCommon.domain.Account;
using AccountDto = tripsGrpc.Account;
using Booking = TripCommon.domain.Booking;
using Trip = TripCommon.domain.Trip;
using TripDto = tripsGrpc.Trip;

namespace Common.Net
{
    public class ProtoUtils
    {

        public static Response CreateLoginResponse(Account account)
        {
            Response response = null;
            if (account == null)
            {
                response = new Response()
                {
                    Type = Response.Types.ResponseType.Error,
                    Error = "Invalid name or password"
                };
            }
            else
            {
                AccountDto a = new AccountDto()
                {
                    Id = account.Id,
                    Name = account.Name,
                    Password = account.Password
                };
                response = new Response()
                {
                    Type = Response.Types.ResponseType.Ok,
                    Account = a
                };
            }
            return response;
        }
        public static Response CreateOkResponse()
        {
            return new Response()
            {
                Type = Response.Types.ResponseType.Ok
            };
        }

        public static Response CreateErrorResponse(Exception exception)
        {
            return new Response()
            {
                Type = Response.Types.ResponseType.Error,
                Error = exception.Message
            };
        }

        public static Response CreateGetTripsResponse(List<Trip> trips)
        {
            Response response = new Response()
            {
                Type = Response.Types.ResponseType.Ok
            };
            trips.ForEach(el =>
            {
                double departure = el.DepartureTime.Subtract(new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc)).TotalMilliseconds;
                TripDto trip = new TripDto()
                {
                    Id = el.Id,
                    Landmark = el.Landmark,
                    CompanyName = el.CompanyName,
                    DepartureTime = departure,
                    Price = el.Price,
                    AvailablePlaces = el.AvailablePlaces
                };
                response.Trips.Add(trip);
            });
            return response;
        }

        public static Trip ConvertToTrip(TripDto t)
        {
            var posixTime = DateTime.SpecifyKind(new DateTime(1970, 1, 1, 0, 0, 0), DateTimeKind.Utc);
            DateTime departure = posixTime.AddMilliseconds(t.DepartureTime);
            return new Trip(t.Id, t.Landmark, t.CompanyName, departure, t.Price, t.AvailablePlaces);
        }

        public static Account ConvertToAccount(AccountDto account)
        {
            return new Account(account.Id, account.Name, account.Password);
        }

       
        public static Account GetAccount(Request request)
        {
            AccountDto a = request.Account;
            return new Account(a.Id, a.Name, a.Password);
        }

    }
}

File: year2/semester2/mpp/java/TripClientServerGrpc/TripCommon/src/main/java/net/ProtoUtils.java
package net;

import domain.Account;
import domain.Trip;
import tripsGrpc.Booking;
import tripsGrpc.Request;
import tripsGrpc.Response;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.List;
import java.util.stream.Collectors;

public class ProtoUtils {
    public static Request createLoginRequest(String name, String password) {
        tripsGrpc.Account account = tripsGrpc.Account.newBuilder().setName(name).setPassword(password).build();
        return Request.newBuilder().setAccount(account).build();
    }

    public static Request createRequest() {
        return Request.newBuilder().build();
    }

    public static Request createSearchTripsRequest(String landmark, int start, int end) {
        tripsGrpc.Trip tripDto = tripsGrpc.Trip.newBuilder().setLandmark(landmark).setStart(start).setEnd(end).build();
        return Request.newBuilder().setTrip(tripDto).build();
    }

    public static Request createAddBookingRequest(String client, String phone, int numTickets, Trip trip, Account account) {
        LocalDateTime dep = trip.getDepartureTime();
        Instant instant = dep.atZone(ZoneId.of("UTC")).toInstant();
        double timeInMillis = instant.toEpochMilli();
        tripsGrpc.Trip t = tripsGrpc.Trip.newBuilder().setId(trip.getId()).setLandmark(trip.getLandmark())
                .setAvailablePlaces(trip.getAvailablePlaces()).setCompanyName(trip.getCompanyName())
                .setPrice(trip.getPrice()).setDepartureTime(timeInMillis).build();
        tripsGrpc.Account a = tripsGrpc.Account.newBuilder().setId(account.getId()).setName(account.getName())
                .setPassword(account.getPassword()).build();
        Booking booking = Booking.newBuilder().setClient(client).setAccount(a).setNumTickets(numTickets)
                .setPhone(phone).setTrip(t).build();
        return Request.newBuilder().setBooking(booking).build();
    }


    public static Account getAccount(Response response) {
        tripsGrpc.Account account = response.getAccount();
        return new Account(account.getId(), account.getName(), account.getPassword());
    }


    public static List<Trip> getTrips(Response response) {
        List<tripsGrpc.Trip> trips = response.getTripsList();
        return trips.stream()
                .map(ProtoUtils::convertToTrip)
                .collect(Collectors.toList());
    }

    public static Trip convertToTrip(tripsGrpc.Trip t) {
        LocalDateTime departure = LocalDateTime.ofInstant(Instant.ofEpochMilli((long) t.getDepartureTime()), ZoneId.of("UTC"));
        return new Trip(t.getId(), t.getLandmark(), t.getCompanyName(), departure, t.getPrice(), t.getAvailablePlaces());
    }

    public static Account convertToAccount(tripsGrpc.Account account) {
        return new Account(account.getId(), account.getName(), account.getPassword());
    }
}
